lib/longtail_platform.c

```diff
@@ -1066,19 +1066,30 @@ int Longtail_GetFilePermissions(const char* path, uint16_t* out_permissions)
     return 0;
 }
 
+#define LONGTAIL_WIN32_PART_SIZE(cur, end) \
+    ((end - cur) > (DWORD)(0xffffffff)) ? (DWORD)(0xffffffff) : (DWORD)(end - cur)
+
 int Longtail_Read(HLongtail_OpenFile handle, uint64_t offset, uint64_t length, void* output)
 {
     HANDLE h = (HANDLE)(handle);
 
     OVERLAPPED ReadOp;
     memset(&ReadOp, 0, sizeof(ReadOp));
 
-    ReadOp.Offset  = (DWORD)(offset & 0xffffffff);
-    ReadOp.OffsetHigh = (DWORD)(offset >> 32);
-
-    if (FALSE == ReadFile(h, output, (DWORD)length, 0, &ReadOp))
+    char* cur = (char*)output;
+    const char* end = cur + length;
+    while (cur < end)
     {
-        return Win32ErrorToErrno(GetLastError());
+        ReadOp.Offset  = (DWORD)(offset & 0xffffffff);
+        ReadOp.OffsetHigh = (DWORD)(offset >> 32);
+
+        const DWORD part_size = LONGTAIL_WIN32_PART_SIZE(cur, end);
+        if (FALSE == ReadFile(h, cur, part_size, 0, &ReadOp))
```

Comment:
We should make use of the fourth parameter - `lpNumberOfBytesRead` to determine how many bytes we actually read.

Reply:
Agreed that using the actual bytes read/written is better.

There is some caveat to `lpNumberOfBytesRead` when using the optional `OVERLAPPED` parameter, but I don't remember the details. Will look into that.

Reply:
Updated in latest. The detail is that if the file was opened with `FILE_FLAG_OVERLAPPED` then `lpNumberOfBytesRead` must be NULL.